spv/ancestors_binary.go

```diff
@@ -0,0 +1,268 @@
+package spv
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/libsv/go-bk/crypto"
+	"github.com/libsv/go-bt/v2"
+	"github.com/libsv/go-bt/v2/bscript"
+	"github.com/pkg/errors"
+
+	"github.com/libsv/go-bc"
+)
+
+const (
+	flagTx    = byte(1)
+	flagProof = byte(2)
+	flagMapi  = byte(3)
+)
+
+// Ancestry is a payment and its ancestors.
+type Ancestry struct {
+	PaymentTx *bt.Tx
+	Ancestors map[[32]byte]*Ancestor
+}
+
+// Ancestor is an internal struct for validating transactions with their ancestors.
+type Ancestor struct {
+	Tx            *bt.Tx
+	Proof         []byte
+	MapiResponses []*bc.MapiCallback
+}
+
+// binaryChunk is a clear way to pass around chunks while keeping their type.
+type binaryChunk struct {
+	ContentType byte
+	Data        []byte
+}
+
+type extendedInput struct {
+	input *bt.Input
+	vin   int
+}
+
+// NewAncestryFromBytes creates a new struct from the bytes of a txContext.
+func NewAncestryFromBytes(b []byte) (*Ancestry, error) {
+	offset := uint64(1)
+	total := uint64(len(b))
+
+	l, size := bt.DecodeVarInt(b[offset:])
+	offset += uint64(size)
+	paymentTx, err := bt.NewTxFromBytes(b[offset : offset+l])
+	if err != nil {
+		return nil, err
+	}
+	ancestry := &Ancestry{
+		PaymentTx: paymentTx,
+		Ancestors: make(map[[32]byte]*Ancestor),
+	}
+	offset += l
+
+	var TxID [32]byte
+
+	if total == offset {
+		return nil, ErrCannotCalculateFeePaid
+	}
+
+	// You're not allowed to just have payment tx with a proof.
+	if b[offset] != 1 {
+		return nil, ErrTipTxConfirmed
+	}
+
+	for total > offset {
+		chunk := parseChunk(b, &offset)
+		switch chunk.ContentType {
+		case flagTx:
+			hash := crypto.Sha256d(chunk.Data)
+			copy(TxID[:], bt.ReverseBytes(hash)) // fixed size array from slice.
+			tx, err := bt.NewTxFromBytes(chunk.Data)
+			if err != nil {
+				return nil, err
+			}
+			if len(tx.Inputs) == 0 {
+				return nil, ErrNoTxInputsToVerify
+			}
+			ancestry.Ancestors[TxID] = &Ancestor{
+				Tx: tx,
+			}
+		case flagProof:
+			ancestry.Ancestors[TxID].Proof = chunk.Data
+		case flagMapi:
+			callBacks, err := parseMapiCallbacks(chunk.Data)
+			if err != nil {
+				return nil, err
+			}
+			ancestry.Ancestors[TxID].MapiResponses = callBacks
+		default:
+			continue
+		}
+	}
+	return ancestry, nil
+}
+
+func parseChunk(b []byte, offset *uint64) binaryChunk {
```

Comment:
NOTE: This is only an opinion.

While the approach of passing a pointer to offset to modify its value is fine and works 100%, it has an implicit uncommon behaviour of the offset being modified that isn't immediately obvious when reading the code.
A clearer approach that actually closer matches the standard lib does would be to have the function return the bytes it was offset by, and the caller of the function increments the offset by that amount.

So the signature of this func would be:
```
func parseChunk(b []byte, offset uint64) (binaryChunky, uint64)
```

Internally, a `bytesRead` (or any name really) would be called up, looking something like:
```
	var bytesRead uint64

	typeOfNextData := b[offset]
	bytesRead++

	l, size := bt.DecodeVarInt(b[offset+bytesRead:])

	bytesRead += uint64(size)
	chunk := binaryChunk{
		ContentType: typeOfNextData,
		Data:        b[offset+bytesRead : offset+bytesRead+l],
	}

	return chunk, byteRead + l
```

And on call:
```
chunk, n := parseChunk(b, offset)
offset += n
```

Reply:
Beautiful, this is much cleaner - thanks for the suggestion.

Reply:
done, or at least very similar solution done.